daemon/disk_usage.go

```diff
@@ -58,77 +72,96 @@ func (daemon *Daemon) imageDiskUsage(ctx context.Context) ([]*image.Summary, err
 
 // localVolumesSize obtains information about volume disk usage from volumes service
 // and makes sure that only one size calculation is performed at the same time.
-func (daemon *Daemon) localVolumesSize(ctx context.Context) ([]*volume.Volume, error) {
-	volumes, _, err := daemon.usageVolumes.Do(ctx, struct{}{}, func(ctx context.Context) ([]*volume.Volume, error) {
+func (daemon *Daemon) localVolumesSize(ctx context.Context) (*types.VolumeDiskUsage, error) {
+	volumes, _, err := daemon.usageVolumes.Do(ctx, struct{}{}, func(ctx context.Context) (*types.VolumeDiskUsage, error) {
 		volumes, err := daemon.volumes.LocalVolumesSize(ctx)
 		if err != nil {
 			return nil, err
 		}
-		return volumes, nil
+
+		du := &types.VolumeDiskUsage{Items: volumes}
+		for _, v := range du.Items {
+			if v.UsageData.Size != -1 {
+				if v.UsageData.RefCount == 0 {
+					du.Reclaimable += v.UsageData.Size
+				}
+				du.TotalSize += v.UsageData.Size
+			}
+		}
+		return du, nil
 	})
 	return volumes, err
 }
 
+type layerDiskUsage struct {
+	LayersSize  int64
+	ContentSize int64
+}
+
 // layerDiskUsage obtains information about layer disk usage from image service
 // and makes sure that only one size calculation is performed at the same time.
 func (daemon *Daemon) layerDiskUsage(ctx context.Context) (int64, error) {
-	usage, _, err := daemon.usageLayer.Do(ctx, struct{}{}, func(ctx context.Context) (int64, error) {
-		usage, err := daemon.imageService.LayerDiskUsage(ctx)
-		if err != nil {
-			return 0, err
-		}
-		return usage, nil
+	usage, _, err := daemon.usageLayer.Do(ctx, struct{}{}, func(ctx context.Context) (usage int64, err error) {
+		return daemon.imageService.ImageDiskUsage(ctx)
 	})
 	return usage, err
 }
 
 // SystemDiskUsage returns information about the daemon data disk usage.
 // Callers must not mutate contents of the returned fields.
-func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsageOptions) (*types.DiskUsage, error) {
+func (daemon *Daemon) SystemDiskUsage(ctx context.Context, opts system.DiskUsageOptions) (du *types.DiskUsage, err error) {
```

Comment:
Try to avoid named output variables; I don't think we need them in this case. We try to avoid them and only use them in exceptional cases (usually if there's an error to be handled in a `defer`, and in that case we make sure it's not named `err`, but (e.g.) `retErr` to make it less err-prone (we've had various cases where `err` was accidentally shadowed).